packages/core/aqua-wrapper/src/generate/service.ts

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2023 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ServiceDef } from '@fluencelabs/interfaces';
+import { recursiveRenameLaquaProps } from '../utils.js';
+import { TypeGenerator } from './interfaces.js';
+
+interface DefaultServiceId {
+    s_Some__f_value?: string
+}
+
+export class ServiceGenerator {
```

Comment:
since `ServiceGenerator` doesn't have any state and is never reused and in fact is just a pure function - why compliate things with classes? I would suggest just converting this to a function `generate` that would also accept `outputType: OutputType` as parameter
TSTypeGenerator and JSTypeGenerator - are just a bunch of pure functions as well so if you just create a simple object litteral

```typescript
const typeGenerators: {[OutputType]: TypeGenerator} = { ... }
```

You will be able to just use the typeGenerator that you need inside the pure function like this
```typescript
const typeGenerator = typeGenerators[outputType]
```
without needlessly creating objects each time

Reply:
These classes not a simple functions - they instances of EntityGenerator interface (which does not exists currently for simplicity reason). For now we creating services and functions - 2 entities. Later we could have 3 or 4 entities. Entity could have state. And we would generate code with [EntityA, EntityB].map(e => e.generate()) where generate is interface function. Also ServiceGenerator uses `typeGenerator` in 3 class methods - passing them through func param would be cumbersome.

Reply:
Right now these classes can be reduced to pure functions and it would be simpler if they are in fact pure functions. I don't really see any state appearing in the future. If for some unknown reason it appears in the future - there are ways to solve this as well. I believe there shouldn't be anything impure here - these are just transformations.

Having one more arg in function is not that bad as having extra wrappers and useless objects generated for each file we compile

Reply:
Pureness is not a point in making decision between function or class approach. Pure functions could be 5 000 lines long and write them only in functions sounds strange.
You can rewrite every class in function if you move all 20 constructor params to function signature. When this would be good and when not so good? I think it depends.

Simplicity is a good point. Thats why we don't create separate class for util functions for example.

I feel like classes can be used if you have some kind of hierarchy or connection between classes. For example i use as a dependency of ServiceGenerator polymorphic type - TypeGenerator. These classes have a connection. Personally i like express these connections in classes.

Reply:
With my suggestion you will still have this connection - a type called TypeGenerator that you will have to "implement" by creating an object literal. It's just that you will have to do it once. Having static class would be an equivalent way of grouping some functions.

So my two, I hope, somewhat objective, points are:
- pure functions are simpler then classes
- you are needlessly creating some garbage for v8 to remove by creating one-off objects for each file that is compiled (I notice this can be metigated if you pass already instantiated object down and just use it's method effectively as a pure function) Just to be clear - I don't think this will have a perfomance hit, or matter as much if leave it as it is (at least at our current scale, because nobody compiles thousands of aqua files) but it's just worse design in my opinion, that I can't approve

Reply:
Don't know why you decided to do only half of the refactoring
As I said, I will refrase, instead of passing down `TypeGenerator` you could'v passed down only `OutputType` instead and access a literal object using this `OutputType`
```
const typeGenerators: Record<OutputType, TypeGenerator> = { ... }
const typeGenerator = typeGenerators[outputType]
```
You will not loose any of your guarantees and you will effectivly have a singleton for this gorup of funtions